### PR TITLE
[MLIR] fix shape.broadcast canonicalize with all empty shape operands

### PR DESCRIPTION
**Pull request by YellowHCH**
Example: all the operands of `shape.broadcast` are empty tensors.
```
func.func @all_empty(%arg0: tensor<f32>) -> tensor<0xindex> {
  %1 = shape.shape_of %arg0 : tensor<f32> -> tensor<0xindex>
  %2 = shape.const_shape [] : tensor<0xindex>
  %3 = shape.broadcast %1, %2, %1 : tensor<0xindex>, tensor<0xindex>, tensor<0xindex> -> tensor<0xindex>
  return %3 : tensor<0xindex>
}
```

One can reproduce crash when canonicalize with *down-top* order, cmd like this:
`mlir-opt -split-input-file -allow-unregistered-dialect -canonicalize="test-convergence top-down=0" %s`

The root cause is when all operands are empty tensor, `RemoveEmptyShapeOperandsPattern` would filter out all operands.

### CODE REVIEW

File: mlir/lib/Dialect/Shape/IR/Shape.cpp [hunk 1]
```diff
@@ -699,6 +699,12 @@ struct RemoveEmptyShapeOperandsPattern : public OpRewritePattern<OpTy> {
                                                  isPotentiallyNonEmptyShape);
 
     // Reduce op to equivalent without empty shape operands.
```
Comment (Lewuathe) [inline mlir/lib/Dialect/Shape/IR/Shape.cpp:701]: I think this comment is more applicable to the next branch starting from L708? 

The new code seems to replace the op with **the empty shape operands**. The comment can be something like this?

```
// Replace the op with empty shape constant if all operants are reduced to be empty. 
```

File: mlir/test/Dialect/Shape/canonicalize.mlir [hunk 2]
```diff
@@ -1,4 +1,5 @@
 // RUN: mlir-opt -split-input-file -allow-unregistered-dialect -canonicalize="test-convergence" %s | FileCheck %s
+// RUN: mlir-opt -split-input-file -allow-unregistered-dialect -canonicalize="test-convergence top-down=1" %s | FileCheck %s
```
Comment (Lewuathe) [inline mlir/test/Dialect/Shape/canonicalize.mlir:2]: In the original issue, we use `top-down=0` to reproduce the case. Is it okay to use `top-down=1` instead?

https://github.com/llvm/llvm-project/pull/118941

Comment (YellowHCH) [reply mlir/test/Dialect/Shape/canonicalize.mlir]: It should be `top-down=0`. Thanks for your reminder.

**Review by Lewuathe — APPROVED** [review]
It looks good to me now. Thanks for updating!

**Pull request merged**